File: 백준/Gold/2629. 양팔저울/양팔저울.py
import sys
import copy
from collections import deque

input = sys.stdin.readline

n = int(input())
arr1 = list(map(int, input().split()))
m = int(input())
arr2 = list(map(int, input().split()))

dp = [[0 for j in range((i+1)*500 +1)] for i in range(n+1)]
ans = []

def sol(i, target):
    if i > n:
        return
    if dp[i][target]:
        return

    dp[i][target] = 1

    sol(i+1, target)
    sol(i+1, target + arr1[i-1])
    sol(i+1, abs(target - arr1[i-1]))
    
sol(0,0)

for i in arr2:
    if i > 500*30:
        ans.append("N")
    elif dp[n][i] == 1:
        ans.append("Y")
    else:
        ans.append("N")

print(" ".join(ans))

File: 백준/Gold/2629. 양팔저울/양팔저울.go
package main

import (
	"bufio"
	"os"
	"strconv"
	"strings"
)

const maxWeight = 500

var (
	dp   [][]bool
	n    int
	arr1 []int
	arr2 []int
)

func sol(i, target int) {
	if i > n {
		return
	}
	if dp[i][target] {
		return
	}

	dp[i][target] = true

	sol(i+1, target)
	if i < n {
		sol(i+1, target+arr1[i])
		sol(i+1, abs(target-arr1[i]))
	}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	line, _ := reader.ReadString('\n')
	n, _ = strconv.Atoi(strings.TrimSpace(line))

	line, _ = reader.ReadString('\n')
	arr1 = toIntArray(strings.Fields(line))

	reader.ReadString('\n')
	line, _ = reader.ReadString('\n')
	arr2 = toIntArray(strings.Fields(line))

	dp = make([][]bool, n+1)
	for i := 0; i <= n; i++ {
		dp[i] = make([]bool, maxWeight*30+1)
	}

	sol(0, 0)

	for _, weight := range arr2 {
		if weight > maxWeight*30 {
			writer.WriteString("N ")
		} else if dp[n][weight] {
			writer.WriteString("Y ")
		} else {
			writer.WriteString("N ")
		}
	}
}

func toIntArray(strs []string) []int {
	arr := make([]int, len(strs))
	for i, s := range strs {
		arr[i], _ = strconv.Atoi(s)
	}
	return arr
}
